File: cli-creator/examples/python-crud-cli/output.py
"""
Output layer: handles --json, --raw, and default TTY-aware formatting.
"""

from __future__ import annotations

import json
import sys
from typing import Any


def print_output(
    data: Any,
    *,
    json_mode: bool = False,
    raw: bool = False,
) -> None:
    """Print data to stdout in the appropriate format.

    - raw: pass through exactly as received
    - json_mode: guaranteed parseable JSON (wraps strings in {"text": "..."})
    - default: JSON pretty-print in TTY, clean JSON in pipes
    """
    if raw:
        print(data if isinstance(data, str) else json.dumps(data))
        return

    if json_mode or not sys.stdout.isatty():
        if isinstance(data, str):
            output = json.dumps({"text": data}, indent=2)
        else:
            output = json.dumps(data, indent=2, default=str)
        print(output)
        return

    # Human-friendly TTY output
    _print_pretty(data)


def _print_pretty(data: Any) -> None:
    """Print human-friendly output to the terminal."""
    if isinstance(data, list):
        _print_table(data)
    elif isinstance(data, dict):
        _print_key_value(data)
    else:
        print(data)


def _print_table(rows: list[Any]) -> None:
    """Print a list of dicts as a simple table."""
    if not rows:
        print("(no results)")
        return

    first = rows[0]
    if not isinstance(first, dict):
        for row in rows:
            print(row)
        return

    keys = list(first.keys())
    widths = [
        max(len(k), *(len(str(row.get(k, ""))) for row in rows))
        for k in keys
    ]

    # Header
    print("  ".join(k.ljust(w) for k, w in zip(keys, widths)))
    print("  ".join("─" * w for w in widths))

    # Rows
    for row in rows:
        print("  ".join(str(row.get(k, "")).ljust(w) for k, w in zip(keys, widths)))


def _print_key_value(obj: dict[str, Any]) -> None:
    """Print a dict as aligned key-value pairs."""
    if not obj:
        print("{}")
        return

    max_key_len = max(len(k) for k in obj)
    for key, value in obj.items():
        display = json.dumps(value) if isinstance(value, (dict, list)) else str(value)
        print(f"{key.ljust(max_key_len)}  {display}")

File: cli-creator/examples/node-search-cli/output.ts
/**
 * Output layer: handles --json, --raw, and default TTY-aware formatting.
 */

export interface OutputOptions {
  json?: boolean;
  raw?: boolean;
}

/**
 * Print data to stdout in the appropriate format.
 *
 * - --raw: pass through exactly as received
 * - --json: guaranteed parseable JSON (wraps strings in { text: "..." })
 * - default: JSON pretty-print in TTY, clean JSON in pipes
 */
export function printOutput(data: unknown, opts: OutputOptions = {}): void {
  if (opts.raw) {
    process.stdout.write(
      typeof data === "string" ? data : JSON.stringify(data)
    );
    return;
  }

  if (opts.json || !process.stdout.isTTY) {
    const output =
      typeof data === "string"
        ? JSON.stringify({ text: data }, null, 2)
        : JSON.stringify(data, null, 2);
    console.log(output);
    return;
  }

  // Human-friendly TTY output
  printPretty(data);
}

/**
 * Print human-friendly output to the terminal.
 */
function printPretty(data: unknown): void {
  if (Array.isArray(data)) {
    printTable(data);
  } else if (typeof data === "object" && data !== null) {
    printKeyValue(data as Record<string, unknown>);
  } else {
    console.log(data);
  }
}

/**
 * Print an array of objects as a simple table.
 */
function printTable(rows: unknown[]): void {
  if (rows.length === 0) {
    console.log("(no results)");
    return;
  }

  const first = rows[0];
  if (typeof first !== "object" || first === null) {
    rows.forEach((row) => console.log(row));
    return;
  }

  const keys = Object.keys(first);
  const widths = keys.map((k) =>
    Math.max(
      k.length,
      ...rows.map((r) => String((r as Record<string, unknown>)[k] ?? "").length)
    )
  );

  // Header
  console.log(keys.map((k, i) => k.padEnd(widths[i])).join("  "));
  console.log(widths.map((w) => "─".repeat(w)).join("  "));

  // Rows
  for (const row of rows) {
    const r = row as Record<string, unknown>;
    console.log(keys.map((k, i) => String(r[k] ?? "").padEnd(widths[i])).join("  "));
  }
}

/**
 * Print an object as key-value pairs.
 */
function printKeyValue(obj: Record<string, unknown>): void {
  const maxKeyLen = Math.max(...Object.keys(obj).map((k) => k.length));
  for (const [key, value] of Object.entries(obj)) {
    const displayValue =
      typeof value === "object" ? JSON.stringify(value) : String(value);
    console.log(`${key.padEnd(maxKeyLen)}  ${displayValue}`);
  }
}
